### PR TITLE
lvol/blob: add shallow copy over external device

### PR DESCRIPTION
**Pull request by DamiaSan**
A shallow copy will copy over the destination device only the cluster allocated to the blob/lvol discarding those belonging to the blob/lvol parent snapshot

### CODE REVIEW

File: doc/jsonrpc.md [hunk 1]
```diff
@@ -9973,6 +9974,44 @@ Example response:
...
+
+#### Example
+
+Example request:
+
+~~~json
+{
+  "jsonrpc": "2.0",
+  "method": "bdev_lvol_shallow_copy",
+  "id": 1.
```
Comment (derekbit) [inline doc/jsonrpc.md]:
```suggestion
  "id": 1,
```

File: include/spdk/blob.h [hunk 2]
```diff
@@ -676,6 +676,23 @@ void spdk_bs_inflate_blob(struct spdk_blob_store *bs, struct spdk_io_channel *ch
 void spdk_bs_blob_decouple_parent(struct spdk_blob_store *bs, struct spdk_io_channel *channel,
 				  spdk_blob_id blobid, spdk_blob_op_complete cb_fn, void *cb_arg);
 
+/**
+ * Perform a shallow copy over an external device
```
Comment (derekbit) [inline include/spdk/blob.h]: Can we get the progress of the copy task on the fly?

File: test/unit/lib/lvol/lvol.c/lvol_ut.c [hunk 3]
```diff
@@ -120,7 +120,7 @@ spdk_bs_iter_first(struct spdk_blob_store *bs,
 uint64_t
 spdk_blob_get_num_clusters(struct spdk_blob *blob)
 {
-	return 0;
+	return 1;
```
Comment (derekbit) [inline test/unit/lib/lvol/lvol.c/lvol_ut.c]: What's the purpose of the change?

File: test/unit/lib/blob/blob.c/blob_ut.c [hunk 4]
```diff
@@ -7394,6 +7395,127 @@ blob_seek_io_unit(void)
...
+}
+
+static void
+blob_shallow_copy(void)
+{
+	struct spdk_blob_store	*bs = g_bs;
+	struct spdk_blob_opts	blob_opts;
+	struct spdk_blob		*blob;
+	spdk_blob_id			blobid;
+	uint64_t				num_clusters = 4;
```
Comment (derekbit) [inline test/unit/lib/blob/blob.c/blob_ut.c]: Replace the magic number 4 with a meaningful constant.

File: module/bdev/lvol/vbdev_lvol_rpc.c [hunk 5]
```diff
@@ -1092,3 +1092,86 @@ rpc_bdev_lvol_grow_lvstore(struct spdk_jsonrpc_request *request,
...
+	}
+
+	lvol = vbdev_lvol_get_from_bdev(lvol_bdev);
+	if (lvol == NULL) {
+		SPDK_ERRLOG("lvol does not exist\n");
+		spdk_jsonrpc_send_error_response(request, -ENODEV, spdk_strerror(ENODEV));
+		goto cleanup;
+	}
+
+	bdev = spdk_bdev_get_by_name(req.bdev_name);
```
Comment (derekbit) [inline module/bdev/lvol/vbdev_lvol_rpc.c]: nit:
rename to `dst_bdev`

File: module/bdev/lvol/vbdev_lvol_rpc.c [hunk 6]
```diff
@@ -1092,3 +1092,86 @@ rpc_bdev_lvol_grow_lvstore(struct spdk_jsonrpc_request *request,
...
+	if (spdk_json_decode_object(params, rpc_bdev_lvol_shallow_copy_decoders,
+				    SPDK_COUNTOF(rpc_bdev_lvol_shallow_copy_decoders),
+				    &req)) {
+		SPDK_INFOLOG(lvol_rpc, "spdk_json_decode_object failed\n");
+		spdk_jsonrpc_send_error_response(request, SPDK_JSONRPC_ERROR_INTERNAL_ERROR,
+						 "spdk_json_decode_object failed");
+		goto cleanup;
+	}
+
+	lvol_bdev = spdk_bdev_get_by_name(req.lvol_name);
```
Comment (derekbit) [inline module/bdev/lvol/vbdev_lvol_rpc.c]: nit:
rename to `src_bdev`

File: doc/jsonrpc.md [hunk 7]
```diff
@@ -9973,6 +9974,44 @@ Example response:
...
+Make a shallow copy of a logical volume over a bdev external to the logical volume store.
+Only cluster allocated to the logical volume will be written on the bdev.
+
+#### Parameters
+
+Name                    | Optional | Type        | Description
+----------------------- | -------- | ----------- | -----------
+lvol_name               | Required | string      | UUID or alias of the logical volume to create a copy from
+bdev_name               | Required | string      | Name of the bdev that acts as destination for the copy
+
```
Comment (derekbit) [inline doc/jsonrpc.md]: Suggest renaming `src_lvol_name` and `dst_bdev_name`?

File: module/bdev/lvol/vbdev_lvol.c [hunk 8]
```diff
@@ -1709,4 +1709,65 @@ vbdev_lvs_grow(struct spdk_lvol_store *lvs,
...
+
+void
+vbdev_lvol_shallow_copy(struct spdk_lvol *lvol, const char *bdev_name,
+			spdk_lvol_op_complete cb_fn, void *cb_arg)
+{
+	struct spdk_bs_dev *ext_dev;
+	struct spdk_lvol_copy_req *req;
+	int rc;
+
+	if (lvol == NULL) {
```
Comment (derekbit) [inline module/bdev/lvol/vbdev_lvol.c]: Check `bdev_name` as well for the consistency.

File: module/bdev/lvol/vbdev_lvol.c [hunk 9]
```diff
@@ -1709,4 +1709,65 @@ vbdev_lvs_grow(struct spdk_lvol_store *lvs,
 	}
 }
 
+static void
```
Comment (derekbit) [inline module/bdev/lvol/vbdev_lvol.c]: Do we need any error handling or printing logs in the callback function?

File: module/bdev/lvol/vbdev_lvol.c [hunk 10]
```diff
@@ -1709,4 +1709,65 @@ vbdev_lvs_grow(struct spdk_lvol_store *lvs,
...
+			    lvolerrno);
+	}
+
+	req->ext_dev->destroy(req->ext_dev);
+	req->cb_fn(req->cb_arg, lvolerrno);
+	free(req);
+}
+
+void
+vbdev_lvol_shallow_copy(struct spdk_lvol *lvol, const char *bdev_name,
```
Comment (derekbit) [inline module/bdev/lvol/vbdev_lvol.c]: Suggest renaming the paramaters to `src_lvol` and `dst_bdev_name`.

File: lib/blob/blobstore.c [hunk 11]
```diff
@@ -6645,6 +6647,216 @@ spdk_bs_blob_decouple_parent(struct spdk_blob_store *bs, struct spdk_io_channel
...
+	if (blob_block_size % ext_dev->blocklen != 0) {
+		SPDK_ERRLOG("external device block size is not compatible with blobstore block size\n");
+		ctx->bserrno = -EINVAL;
+		spdk_blob_close(_blob, bs_shallow_copy_cleanup_finish, ctx);
+		return;
+	}
+
+	ctx->blob = _blob;
+
+	if (_blob->locked_operation_in_progress) {
```
Comment (derekbit) [inline lib/blob/blobstore.c]: Will the blob copy be retried internally? Or, the whole shallow copy will error out?

Comment (derekbit) [reply lib/blob/blobstore.c]: Another question: is `_blob` a source block, right? If yes, the operation is `read`, should we always reject the operation? Any readwrite lock for the blob read access?

File: test/unit/lib/blob/blob.c/blob_ut.c (hunk 4 above)
Comment (derekbit) [reply test/unit/lib/blob/blob.c/blob_ut.c]: Why 4?

File: module/bdev/lvol/vbdev_lvol.c [hunk 12]
```diff
@@ -1709,4 +1709,65 @@ vbdev_lvs_grow(struct spdk_lvol_store *lvs,
...
+	if (lvol == NULL) {
+		SPDK_ERRLOG("lvol does not exist\n");
+		cb_fn(cb_arg, -EINVAL);
+		return;
+	}
+
+	assert(lvol->bdev != NULL);
+
+	req = calloc(1, sizeof(*req));
+	if (req == NULL) {
```
Comment (shuo-wu) [inline module/bdev/lvol/vbdev_lvol.c]: NIT: Add `SPDK_ERRLOG("Cannot alloc memory for vbdev lvol store request pointer\n");`

File: doc/jsonrpc.md [hunk 13]
```diff
@@ -9973,6 +9974,44 @@ Example response:
 }
 ~~~
 
+### bdev_lvol_shallow_copy {#rpc_bdev_lvol_shallow_copy}
+
+Make a shallow copy of a logical volume over a bdev external to the logical volume store.
```
Comment (shuo-wu) [inline doc/jsonrpc.md]: Actually, this statement is a little bit confusing...

File: module/bdev/lvol/vbdev_lvol.c (hunk 9 above)
Comment (shuo-wu) [reply module/bdev/lvol/vbdev_lvol.c]: Similar question. And do we need to add a new `spdk_bdev_event_type`?

File: test/unit/lib/lvol/lvol.c/lvol_ut.c [hunk 14]
```diff
@@ -2089,6 +2097,69 @@ lvol_get_xattr(void)
...
 
+static void
+lvol_shallow_copy(void)
+{
+	struct lvol_ut_bs_dev bs_dev;
+	struct spdk_lvs_opts opts;
+	struct spdk_bs_dev ext_dev;
+	int rc = 0;
+
+	init_dev(&bs_dev);
```
Comment (derekbit) [inline test/unit/lib/lvol/lvol.c/lvol_ut.c]: add `init_dev(&ext_dev);`

File: doc/jsonrpc.md [hunk 15]
```diff
@@ -9973,6 +9974,44 @@ Example response:
 }
 ~~~
 
+### bdev_lvol_shallow_copy {#rpc_bdev_lvol_shallow_copy}
```
Comment (derekbit) [inline doc/jsonrpc.md]: Should we need `bdev_lvol_shallow_copy_status`? The response contains `status` and `progress`.

File: lib/lvol/lvol.c [hunk 16]
```diff
@@ -1521,3 +1521,72 @@ spdk_lvs_grow(struct spdk_bs_dev *bs_dev, spdk_lvs_op_with_handle_complete cb_fn
...
+	req->cb_arg = cb_arg;
+	req->channel = spdk_bs_alloc_io_channel(lvol->lvol_store->blobstore);
+	if (req->channel == NULL) {
+		SPDK_ERRLOG("Cannot alloc io channel for lvol shallow copy request\n");
+		free(req);
+		cb_fn(cb_arg, -ENOMEM);
+		return;
+	}
+
+	blob_id = spdk_blob_get_id(lvol->blob);
```
Comment (shuo-wu) [inline lib/lvol/lvol.c]: Should we validate if the source lvol is snapshot/read_only?

File: lib/blob/blobstore.c [hunk 17]
```diff
@@ -6645,6 +6647,216 @@ spdk_bs_blob_decouple_parent(struct spdk_blob_store *bs, struct spdk_io_channel
...
+		SPDK_ERRLOG("Shallow copy blob open error %d\n", bserrno);
+		ctx->bserrno = bserrno;
+		bs_shallow_copy_cleanup_finish(ctx, bserrno);
+		return;
+	}
+
+	blob_block_size = _blob->bs->dev->blocklen;
+	blob_total_size = spdk_blob_get_num_clusters(_blob) * spdk_bs_get_cluster_size(_blob->bs);
+
+	if (blob_total_size > ext_dev->blockcnt * ext_dev->blocklen) {
```
Comment (shuo-wu) [inline lib/blob/blobstore.c]: Should we consider the apparent size issue?
For example, let's say the source lvol contains 1 cluster only and the cluster size is 4MB. Hence `blob_total_size` is 4MB here. But the offset of the data is 10Gi and the destination block device is size 1Gi. In this case, Does this check still work?

File: lib/blob/blobstore.c (hunk 11 above)
Comment (shuo-wu) [reply lib/blob/blobstore.c]: Here I am thinking if we can bypass this locking by only allowing read-only lvols as sources. But we still need to consider the deletion/decoupling/... cases.

File: lib/blob/blobstore.c [hunk 18]
```diff
@@ -6645,6 +6647,216 @@ spdk_bs_blob_decouple_parent(struct spdk_blob_store *bs, struct spdk_io_channel
...
+	ctx->blob = _blob;
+
+	if (_blob->locked_operation_in_progress) {
+		SPDK_DEBUGLOG(blob, "Cannot make a shallow copy of blob - another operation in progress\n");
+		ctx->bserrno = -EBUSY;
+		spdk_blob_close(_blob, bs_shallow_copy_cleanup_finish, ctx);
+		return;
+	}
+
+	_blob->locked_operation_in_progress = true;
```
Comment (shuo-wu) [inline lib/blob/blobstore.c]: Here, I am curious how SPDK guarantees the thread-safe of this locking.

File: lib/blob/blobstore.c (hunk 17 above)
Comment (DamiaSan) [reply lib/blob/blobstore.c]: `spdk_blob_get_num_clusters` returns the total number of clusters in the volume, not only the allocated ones in case of a tiny provisioned volume.

File: lib/blob/blobstore.c (hunk 11 above)
Comment (DamiaSan) [reply lib/blob/blobstore.c]: > Will the blob copy be retried internally? Or, the whole shallow copy will error out?


The whole shallow copy will error out. Because of the internal structure of the copy, blob store's block size must be a multiple of external device's block size.

Comment (DamiaSan) [reply lib/blob/blobstore.c]: > Here I am thinking if we can bypass this locking by only allowing read-only lvols as sources. But we still need to consider the deletion/decoupling/... cases.

Good idea to let shallow copy only from snapshots, I will insert a check over the lvol

File: lib/blob/blobstore.c (hunk 18 above)
Comment (DamiaSan) [reply lib/blob/blobstore.c]: Only one thread works over metadata because blobstore is designed to be lock-free

File: module/bdev/lvol/vbdev_lvol.c (hunk 9 above)
Comment (DamiaSan) [reply module/bdev/lvol/vbdev_lvol.c]:
```
/** Asynchronous event type */
enum spdk_bdev_event_type {
	SPDK_BDEV_EVENT_REMOVE,
	SPDK_BDEV_EVENT_RESIZE,
	SPDK_BDEV_EVENT_MEDIA_MANAGEMENT,
};
```
What events should we handle?
In case of bdev removal we could stop the shallow copy: to do that I think must insert some kind of synchronization that can be be useful for retrieving shallow copy status too

File: test/unit/lib/blob/blob.c/blob_ut.c (hunk 4 above)
Comment (DamiaSan) [reply test/unit/lib/blob/blob.c/blob_ut.c]: 4 to have a couple of cluster allocated per snapshot interleaved with non allocated clusters.
If you follow the operations it is clearer:
- Write on cluster 2 and 4 of blob
- Make a snapshot over blob
- Write on cluster 1 and 3 of blob
- Make a shallow copy of blob over bdev
The shallow copy must copy only cluster 1 and 3 because 2 and 4 are allocated in the old snapshot

Comment (DamiaSan) [reply test/unit/lib/blob/blob.c/blob_ut.c]: > Replace the magic number 4 with a meaningful constant.

It is a unit test, so if you follow the flow of the operations I think it is clear what it means. Because it is difficult to give it a name ...

File: test/unit/lib/lvol/lvol.c/lvol_ut.c (hunk 14 above)
Comment (DamiaSan) [reply test/unit/lib/lvol/lvol.c/lvol_ut.c]: `ext_dev` is a different type of `bs_dev`, cannot use init_dev

File: test/unit/lib/lvol/lvol.c/lvol_ut.c (hunk 3 above)
Comment (DamiaSan) [reply test/unit/lib/lvol/lvol.c/lvol_ut.c]: The functions existed before but it wasn't used. It is used by `spdk_lvol_shallow_copy` to check blob and bdev size

File: lib/blob/blobstore.c (hunk 18 above)
Comment (shuo-wu) [reply lib/blob/blobstore.c]: Do you mean that, these functions are always handled by a fixed thread? (This issue is unrelated to the PR. I am just curious about the SPDK implementation.)

File: module/bdev/lvol/vbdev_lvol.c (hunk 9 above)
Comment (shuo-wu) [reply module/bdev/lvol/vbdev_lvol.c]: I mean, do we need to add a new event type?

File: lib/blob/blobstore.c (hunk 11 above)
Comment (DamiaSan) [reply lib/blob/blobstore.c]: > Here I am thinking if we can bypass this locking by only allowing read-only lvols as sources. But we still need to consider the deletion/decoupling/... cases.

Yes you are right, other operations in progress can be snapshot, clone, inflate, ... and cannot be executed during a shallow copy because they can change the cluster map.

File: doc/jsonrpc.md (hunk 7 above)
Comment (l-mb) [reply doc/jsonrpc.md]: Yes, I think making it easy to tell apart src and destination is a good idea.

File: lib/blob/blobstore.c (hunk 18 above)
Comment (DamiaSan) [reply lib/blob/blobstore.c]: You can read something [here](https://spdk.io/doc/blob.html#blob_pg_theory)
A further development in shallow copy will be to introduce qos, so to make lighter the load of this thread. Because shallow copy, as inflate and decouple operations, are rpc commands that involve data.

File: include/spdk/blob.h (hunk 2 above)
Comment (DamiaSan) [reply include/spdk/blob.h]: Done, rpc bdev_lvol_shallow_copy_status

File: doc/jsonrpc.md (hunk 15 above)
Comment (DamiaSan) [reply doc/jsonrpc.md]: Done, rpc bdev_lvol_shallow_copy_status

**Comment by DamiaSan** [comment]
PR update
@shuo-wu @derekbit

**Comment by innobead** [comment]
@DamiaSan There are some conflicts after syncing the latest upstream. Please help resolve them.

**Comment by derekbit** [comment]
@DamiaSan 
Just thinking about a question. After finishing the lvol shallow copy, how to add the lvol to the raid1?
